bob/src/cluster/support_types.rs

```diff
@@ -57,3 +57,76 @@ impl From<RemoteDeleteError> for Error {
         w.error
     }
 }
+
+pub(crate) struct IndexMap {
+    indexes: Vec<usize>,
+}
+
+impl IndexMap {
+    /// Create empty indexes map
+    pub(crate) fn new() -> Self {
+        Self { indexes: vec![] }
+    }
+
+    /// Create with indexes with `false` values
+    pub(crate) fn where_not_exists(data: &[bool]) -> Self {
+        Self {
+            indexes: data
+                .iter()
+                .enumerate()
+                .filter(|(_, f)| !**f)
+                .map(|(i, _)| i)
+                .collect(),
+        }
+    }
+
+    /// Check that indexes are empty
+    pub(crate) fn is_empty(&self) -> bool {
+        self.indexes.is_empty()
+    }
+
+    /// Get number of indexes in the map
+    pub(crate) fn len(&self) -> usize {
+        self.indexes.len()
+    }
+
+    /// Add index
+    pub(crate) fn push(&mut self, item: usize) {
+        debug_assert!(!self.indexes.contains(&item));
+        self.indexes.push(item);
+    }
+
+    /// Collect data with allowed indexes
+    pub(crate) fn collect<T: Clone>(
+        &self,
+        data: &[T]
+    ) -> Vec<T> {
+        let mut result = Vec::with_capacity(self.indexes.len());
+        for &i in &self.indexes {
+            if i < data.len() {
```

Comment:
Here it is expected, that `data` is the original data on top of which this `IndexMap` was built. Thus it should be impossible for `i` to larger or equal than `data.len()`. To enforce this I suggest to use `debug_assert!(i <data.len());` instead of condition